File: src/main/kotlin/com/githubyss/common/kit/util/ClassTypeUtils.java
package com.githubyss.common.kit.util;


import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * ClassTypeUtils
 *
 * @author Ace Yan
 * @github githubyss
 * @createdTime 2022/12/12 16:46:33
 */
public class ClassTypeUtils {

    /** *********************** Properties ****************************** */

    private static final String TAG = "ClassTypeUtils";


    /** ************************* Functions ****************************** */

    public static <T> Class<?> getTypeClassFromParadigm(final IFormatter<T> formatter) {
        Type[] genericInterfaces = formatter.getClass().getGenericInterfaces();
        Type type;
        if (genericInterfaces.length == 1) {
            type = genericInterfaces[0];
        }
        else {
            type = formatter.getClass().getGenericSuperclass();
        }
        type = ((ParameterizedType) type).getActualTypeArguments()[0];
        while (type instanceof ParameterizedType) {
            type = ((ParameterizedType) type).getRawType();
        }
        String className = type.toString();
        if (className.startsWith("class ")) {
            className = className.substring(6);
        }
        else if (className.startsWith("interface ")) {
            className = className.substring(10);
        }
        try {
            return Class.forName(className);
        }
        catch (ClassNotFoundException e) {
            e.printStackTrace();
        }
        return null;
    }

    public static Class<?> getClassFromObject(final Object obj) {
        Class<?> objClass = obj.getClass();
        if (objClass.isAnonymousClass() || objClass.isSynthetic()) {
            Type[] genericInterfaces = objClass.getGenericInterfaces();
            String className;
            if (genericInterfaces.length == 1) {// interface
                Type type = genericInterfaces[0];
                while (type instanceof ParameterizedType) {
                    type = ((ParameterizedType) type).getRawType();
                }
                className = type.toString();
            }
            else {// abstract class or lambda
                Type type = objClass.getGenericSuperclass();
                while (type instanceof ParameterizedType) {
                    type = ((ParameterizedType) type).getRawType();
                }
                className = type.toString();
            }

            if (className.startsWith("class ")) {
                className = className.substring(6);
            }
            else if (className.startsWith("interface ")) {
                className = className.substring(10);
            }
            try {
                return Class.forName(className);
            }
            catch (ClassNotFoundException e) {
                e.printStackTrace();
            }
        }
        return objClass;
    }


    public abstract static class IFormatter<T> {
        public abstract String format(T t);
    }

    public <T> Class<T> getTClass() {
        Type type = getClass().getGenericSuperclass();
        if (!(type instanceof ParameterizedType)) {
            throw new IllegalStateException("Type must be a parameterized type");
        }
        ParameterizedType parameterizedType = (ParameterizedType) type;
        // 获取泛型的具体类型  这里是单泛型
        Type[] actualTypeArguments = parameterizedType.getActualTypeArguments();
        if (null == actualTypeArguments || actualTypeArguments.length < 1) {
            throw new IllegalStateException("Number of type arguments must be 1");
        }
        return (Class<T>) actualTypeArguments[0];
    }
}

File: src/main/kotlin/com/githubyss/common/kit/util/ClassTypeUtils.kt
package com.githubyss.common.kit.util

import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type


/**
 * ClassTypeUtils
 *
 * @author Ace Yan
 * @github githubyss
 * @createdTime 2022/04/15 10:56:17
 */

/** ****************************** Properties ****************************** */

private const val TAG: String = "ClassTypeUtils"


/** ****************************** Functions ****************************** */

/**
 * Get actual type of generic nested
 *
 * - Get type of List<String>: [getClassType(List::class.java, String::class.java)]
 * - Get type of List<Int>: [getClassType(List::class.java, Int::class.javaObjectType)]
 * - Get type of List<List<Int>>: [getClassType(List::class.java, getClassType(List::class.java, Int::class.javaObjectType))]
 * - Get type of Map<String, Int>: [getClassType(Map::class.java, String::class.java, Int::class.javaObjectType)]
 * - Get type of Map<String, List<Int>>: [getClassType(Map::class.java, String::class.java, getClassType(List::class.java, Int::class.javaObjectType))]
 *
 * Kotlin 在使用基本类型时，会根据需要自动装箱（如 Int 在 Jvm 中会自动转成 int 或 Integer），而泛型要求使用的必须是 Object 类型，因此在使用 ParameterizedType 传入基本类型时，需要使用它们的包装类，比如 Int 的就是 Int::class.javaObjectType
 *
 * @param
 * @return
 */
fun getClassType(raw: Class<out Any>, vararg args: Type) = object : ParameterizedType {
    override fun getActualTypeArguments(): Array<out Type> {
        return args
    }

    override fun getRawType(): Type {
        return raw
    }

    override fun getOwnerType(): Type? {
        return null
    }
}

/**
 * Get actual type of generic nested
 *
 * - Get type of List<String>: [getClassType<List<String>>()]
 * - Get type of List<Int>: [getClassType<List<Int>>()]
 * - Get type of List<List<Int>>: [getClassType<List<List<Int>>>()]
 * - Get type of Map<String, Int>: [getClassType<Map<String, Int>>()]
 * - Get type of Map<String, List<Int>>: [getClassType<Map<String, List<Int>>>()]
 *
 * @param
 * @return
 */
inline fun <reified T> getClassType() = object : ClassType<T>() {}.type

abstract class ClassType<T> {
    val type = (javaClass.genericSuperclass as ParameterizedType).actualTypeArguments[0] ?: null
}
